whelktool/scripts/cleanups/2020/11/lxl-2791-fix-broken-holds/lxl-2791-fix-broken-holds.groovy

```diff
@@ -0,0 +1,61 @@
+/**
+ * Fix or remove holds that aren't associated with a bib record.
+ *
+ * See LXL-2791 for more info.
+ */
+
+PrintWriter scheduledForUpdating = getReportWriter("scheduled-updates")
+PrintWriter failedUpdating = getReportWriter("failed-updates")
+PrintWriter scheduledForDeletion = getReportWriter("scheduled-for-deletion")
+PrintWriter failedDeleting = getReportWriter("failed-deleting")
+
+Map holds = [:]
+Map validBibs = [:]
+
+// Turn file with XL id - old hold id - old bib id into a map, keyed by XL id
+new File(scriptDir, "id-oldid-bibid.tsv").each {
+    def (id, oldId, bibId) = it.split(/\t/).collect { it.trim() }
+    holds[id] = ["oldId": oldId, "bibId": bibId]
+}
+
+// Make sure the referenced bib records still exist. Fetch them using the legacy IDs
+// we got from Voyager, and map them to their XL shortIds.
+selectByIds(holds.collect { "http://libris.kb.se/resource/bib/" + it.value.bibId }) { data ->
+    if (data.graph[1]["@type"] == "Instance" && data.graph[0].controlNumber)
```

Comment:
Are we sure that they're all "Instance" ? No "Electronic", "Print" or anything else?

JsonLd.getSubClasses() is a tip, in case you want "any" of these! :)

Reply:
We're not! I removed it - checking seems pointless anyway, since I assume http://libris.kb.se/resource/bib/<id> should always resolve to either a bibliographic record or nothing, and we don't care what type it then happens to be.